src/main/java/org/fisco/bcos/channel/client/Service.java

```diff
@@ -915,6 +936,167 @@ public void run() {
+
+        String content = "";
+
+        HeartBeatParser heartBeatParser = new HeartBeatParser(getUsageChlVersion(ctx));
+        String data = new String(msg.getData());
+        try {
+            BcosHeartbeat bcosHeartbeat = heartBeatParser.decode(data);
+            // logger.trace(" heartbeat packet, heartbeat is {} ", bcosHeartbeat);
+            int heartBeat = bcosHeartbeat.getHeartBeat();
+            content = String.valueOf(heartBeat);
+        } catch (UnsupportedEncodingException e) {
+            logger.error("heartbeat packet cannot be parsed, data is {}", data);
+        } catch (Exception e) {
+            logger.error("heartbeat packet exception, data is {}", data);
+        }
+
+        if ("0".equals(content)) {
+            logger.trace("heartbeat packet，send heartbeat packet back");
+            Message response = new Message();
+
+            response.setSeq(msg.getSeq());
+            response.setResult(0);
+            response.setType((short) 0x13);
+
+            try {
+                response.setData(heartBeatParser.encode("1"));
+            } catch (JsonProcessingException e) {
+                logger.error(" write json failed, message is {} ", e.getMessage());
+                return;
+            }
+
+            ByteBuf out = ctx.alloc().buffer();
+            response.writeHeader(out);
+            response.writeExtra(out);
+
+            ctx.writeAndFlush(out);
+        } else if ("1".equals(content)) {
+            logger.trace("heartbeat response");
+        }
+    }
+
+    public org.fisco.bcos.channel.protocol.Version getUsageChlVersion(ChannelHandlerContext ctx) {
+        SocketChannel socketChannel = (SocketChannel) ctx.channel();
+        String hostAddress = socketChannel.remoteAddress().getAddress().getHostAddress();
+        int port = socketChannel.remoteAddress().getPort();
+
+        String remoteEndPoint = hostAddress + ":" + port;
+        AttributeKey<NodeVersion> attributeKey = AttributeKey.valueOf(remoteEndPoint);
+
+        if (ctx.channel().hasAttr(attributeKey)) {
+            NodeVersion nodeVersion = ctx.channel().attr(attributeKey).get();
+            org.fisco.bcos.channel.protocol.Version nodeHightestChlVersion =
+                    org.fisco.bcos.channel.protocol.Version.convert(
+                            nodeVersion.getHighestSupported());
+
+            org.fisco.bcos.channel.protocol.Version SDKHighestVersion =
+                    org.fisco.bcos.channel.protocol.Version.getHighestSupported();
+
+            logger.trace(
+                    " remote host is {}, node channel version info is {}, node highest version is {}, sdk highest version is {}",
+                    remoteEndPoint,
+                    nodeVersion,
+                    nodeHightestChlVersion.getVersionNumber(),
+                    SDKHighestVersion.getVersionNumber());
+
+            if (nodeHightestChlVersion.getVersionNumber() > SDKHighestVersion.getVersionNumber()) {
```

Comment:
plz give some comments for the selection logic

Reply:
> plz give some comments for the selection logic

This logic is droped